### ISSUE TITLE
Debug output for Perl script containing utf8 characters does not equal output written to file

### ISSUE THREAD

**Issue by daveaiello** [Komodo/KomodoEdit#3784]
### Short Summary

While debugging a Perl script, the text displayed in Debug Output does not always equal the output displayed to standard output or written to a file.  This is possibly due to the strings containing utf8 characters.

### Steps to Reproduce

```
#!/usr/bin/perl -w

print "This is a string that's 43 characters long.\n";
print "Thîs is å string that's 43 charactérs long.\n";

1;
```

### Expected results

```
This is a string that's 43 characters long.
Thîs is å string that's 43 charactérs long.
```

### Actual results

```
This is a string that's 43 characters long.
Thîs is å string that's 43 charactérs lon
```

### Platform Information

*Komodo IDE* 
*version 11.1.1, build 91089* 
*macosx, aka MacOS 10.14.6* 

### Additional Information

 
```
[2019-08-25 15:40:42,225] [ERROR] editorconfig: Invalid filename specified
[2019-08-25 15:40:45,228] [ERROR] console-logger: TypeError: n is undefined (2) in chrome://komodo/content/contrib/commonjs/underscore.js:5
Traceback (most recent call last):
  File "chrome://komodo/content/contrib/commonjs/underscore.js", line 5, in 

```

**Comment by daveaiello** [comment]
Perhaps my example code should have stated that the string is 44 characters long (because the string is 44 characters long if you include the new line character)?

I wrote the example code this way to show that each line of Debug Output appears to be truncated by the same number of characters that have accents or other diacritical marks.

I believe if I had written a third print statement, that output would have continued on the second line of Debug Output, rather than being displayed on the third line.

**Comment by daveaiello** [comment]
The reason I discovered this in the first place is because I am debugging a Perl script being developed to support website localizations. Some languages we are processing with this localization script have multiple characters with diacritical marks on every line of text.

The most important features of Komodo IDE for me are remote debugging and the pane that's used for the variable inspector. Debug Output is not quite as necessary as they are, but I'll need to have a Linux shell window on the screen at the same time to mitigate this bug. So my use of Komodo IDE will certainly be less than optimal.

**Comment by daveaiello** [comment]
I had the opportunity to do some debugging to test the viability of having a bash shell window open to the output file.

I tried both
* `tail -f {output file}` and
* `perl -d {parameters} | tee {output file}`

This is with the Perl code including `$|++;` which normally unbuffers standard output.

Neither of these options resulted in the bash shell output window displaying the Perl script's output in real time.

So the Debug Output display problem inside of KomodoIDE is more of a blocker than I realized.

**Comment by th3coop** [comment]
@daveaiello I have some funky steps for you to attempt.  I THINK I've found the bug but I can no longer reproduce the issue with or without the fix.  In the mean time I'm going to build a nightly for you to try.  I'll have to send it to you directly so when that's ready I'll get you to contact support@activestate.com and ask for me.

Here are the steps I'd like you to try(I'm assuming you're on OSX from your screenshots in the forums:
In your terminal
1. `$ cd /Applications/Komodo\ IDE\ 11.app/Contents/Resources/components`
1. `$ cp koRunTerminal.py koRunTerminal.py.orig`
1. Now open koRunTerminal.py in Komodo, *Ctrl+g* to line 735.
1. Replace the existing `class _DebuggerIOWriter` implementation with the code below.
1. Start Komodo
1. cross fingers
1. reproduce issue

Now, the neat thing with Mac is that it likely won't let you run the program if it's contents have been edited.  You might need to set your system prefs to *Allow apps downloaded from: Anywhere*.

```python
class _DebuggerIOWriter(object):
    _com_interfaces_ = [components.interfaces.koIFile]

    def __init__(self, name, terminal):
        # XXX - WeakReference?
        self.__terminal = terminal
        self.__name = name

    def write(self, data):
        log.debug("_DebuggerIOWriter.write:: data: %r", data)
        encodingServices = components.classes['@activestate.com/koEncodingServices;1'].\
                         getService(components.interfaces.koIEncodingServices);
        if not data:
            # Nothing to add, then nothing to do...
            return
        try:
            data, enc, bom = encodingServices.\
                                getUnicodeEncodedStringUsingOSDefault(data)
            log.debug("_DebuggerIOWriter.write:: enc: %s", enc)
            log.debug("_DebuggerIOWriter.write:: bom: %s", bom)
            l = len(data.encode(enc))
            log.debug("_DebuggerIOWriter.write:: we got the encoding")
        except Exception, ex:
            log.exception(ex);
            l = len(data)
            log.debug("_TerminalReader.run:: we did not get the encoding")
        self.__terminal.addText(l, data, self.__name)
```

**Comment by th3coop** [comment]
@daveaiello, I figured out why I couldn't reproduce the issue.  I was reproducing it on Mac only.  The above patch, though helpful, doesn't fix the issue.

**Comment by daveaiello** [comment]
Not sure what you are saying here.

You believe that this is a MacOS-only problem, right? Does the patch you provided fix this?

When I modify the Python source code for Komodo, I have to force MacOS to execute it. Is there any other way to do this than to "Allow apps downloaded from: Anywhere"? Can I trust the app individually?

**Comment by th3coop** [comment]
> Does the patch you provided fix this?

No.  I said that in my last comment.  You don't need to try the patch.

I'm gonna have to punt this until later when I have more time to devote to bugs.  I have to get back to sprint work.

Note:  After chasing all through the debug code I'm relatively confident that the issue lies in perl5db.pl and how it's getting and sending strings to Komodo.

Other notes:
1. debugger.js get Invoker, doExecute
1. process is hooked up to debuggerPane terminal
1. debugger.js.startDebugging swaps terminal to debugTermhandler using setupDebuggerRedirectIOHandles

Since it's a terminal handler it's just receiving dumps to stdout from the process which is the debugger running.  Since the issue doesn't occur during normal running it must be the perl5db.pl lib.
